tests/fastspeech2/test_modeling_fastspeech2.py

```diff
@@ -0,0 +1,267 @@
+            self.parent.assertEqual(result[i].shape, result[i + 1].shape)
+        # check predicted mel-spectrogram has correct dimension
+        self.parent.assertEqual(result.mel_spectrogram.size(2), self.output_frame_dim)
+
+    def prepare_config_and_inputs_for_common(self):
+        config, input_ids = self.prepare_config_and_inputs()
+        inputs_dict = {"input_ids": input_ids}
+        return config, inputs_dict
+
+
+@require_torch
+class FastSpeech2ModelTest(ModelTesterMixin, unittest.TestCase):
+    all_model_classes = (FastSpeech2Model,) if is_torch_available() else ()
+    test_pruning = False
+    test_headmasking = False
+    test_torchscript = False
+
+    def setUp(self):
+        self.model_tester = FastSpeech2ModelTester(self)
+        self.config_tester = ConfigTester(self, config_class=FastSpeech2Config)
+        # FastSpeech2Config does not have `hidden_size``
+        self.config_tester.create_and_test_config_common_properties = lambda: None
+
+    def test_config(self):
+        self.config_tester.run_common_tests()
+
+    def test_model(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    @unittest.skip(reason="FastSpeech2 does not output attentions")
+    def test_attention_outputs(self):
+        pass
+
+    @unittest.skip(reason="FastSpeech2 does not output attentions")
+    def test_retain_grad_hidden_states_attentions(sef):
+        pass
+
+    @unittest.skip(reason="FastSpeech2 does not output hidden_states")
+    def test_hidden_states_output(self):
+        pass
+
+    @unittest.skip(reason="FastSpeech2 does not accept inputs_embeds")
+    def test_inputs_embeds(self):
+        pass
+
+    @unittest.skip(reason="Feed forward chunking is not implemented")
+    def test_feed_forward_chunking(self):
+        pass
+
+    def test_initialization(self):
+        # TODO
+        pass
+
+    @slow
+    def test_model_from_pretrained(self):
+        model = FastSpeech2Model.from_pretrained("jaketae/fastspeech2-commonvoice")
+        self.assertIsNotNone(model)
+
+
+@require_torch
+@slow
+class FastSpeech2ModelIntegrationTest(unittest.TestCase):
+    @unittest.skipIf(torch_device != "cpu", "cannot make deterministic on GPU")
+    def test_inference_integration(self):
+        model = FastSpeech2Model.from_pretrained("jaketae/fastspeech2-ljspeech")
+        model.to(torch_device)
+        tokenizer = FastSpeech2Tokenizer.from_pretrained("jaketae/fastspeech2-ljspeech")
+        text = ["This is a test sentence."]
+        inputs = tokenizer(text, return_tensors="pt", padding=True).to(torch_device)
+
+        torch.manual_seed(0)
+        np.random.seed(0)
+
+        with torch.no_grad():
+            output = model(**inputs, return_dict=True)
+
+        # mel-spectrogram is too large (123, 80),
+        # so only check top-left 100 elements
+        expected_mel_spectrogram = torch.tensor(
```

Comment:
How is the mel spectrogram transformed to actual audio?

Reply:
Is there a most common "waveform-decoder" that we could add to the modeling file? Do those decoders have weights or not? Think the UserXP would be nicer, if the "waveform-decoder" is added to the modeling file as an additional head. If there are multiple ones, we could maybe just add the best one?

Reply:
I agree that having `FastSpeech2ForWaveGeneration` would be quite nice to have. I think HiFi-GAN would be a sensible choice for a decoder since it's part of `fairseq`'s TTS pipeline. Some questions:

- Should HiFi-GAN code be part of `modeling_fastspeech2.py`?
- Do we still want to prefix HiFi-GAN modules with `FastSpeech2`?
- In the off chance that we want to use HiFi-GAN for other TTS models (e.g., [FastPitch](https://github.com/huggingface/transformers/issues/16349), which is next in line), would we be duplicating code?

cc @anton-l

Reply:
I would answer all there questions above with "Yes" - happy to add all the code in the modeling file

Reply:
Hey @patrickvonplaten, thanks for the feedback. I added HiFi-GAN in the lastest commits. Some questions:

1. **HiFi-GAN Trainability.** Should HiFi-GAN (or `FastSpeech2ForWaveformGeneration`) be trainable? HiFi-GAN is trained just like any normal model, but during inference, the user should call `remove_weight_norm`. For now, I just manually called the function in the `FastSpeech2ForWaveformGeneration.forward()` function, but this might mean that the class is not trainable. Alternatively, I could check if the model is in `eval` mode and only call `remove_weight_norm` if it's in `model.train` is `False`. Do you have a preference here?
2. **HiFi-GAN Configuration.** I hard-coded all HiFi-GAN parameters instead of polluting `FastSpeech2Config`. HiFi-GAN does come in 3 variations (check `config.jsons` [here](https://github.com/jik876/hifi-gan)), but the V1 model has the most amount of checkpoints, covering LJ Speech, VCTK, and a "universal" version for general-purpose decoding. The hard-coded parameters thus follow the V1 configuration.
3. **Waveform Output.** Because input tokens are padded, the output mel-spectrogram will also be of different lengths (although they will appear to be of equal lengths due to batching). When we pass this batch of padded mel-spectrograms to HiFi-GAN, the vocoder will also produce padded audio (again, due to batching, the waveforms may appear to be of equal length, but the shorter samples will contain meaningless noise at the end). I'm trying to think of a good UI/UX to deal with this. Any input would be greatly appreciated!

I realize this PR has been WIP for a very long time. Apologies for the delay, and thank you so much for the review!